File: Tutorials/Codes/Modern_Control_Course/Chapter13/Lesson4/Chapter13_Lesson4.cpp
// Chapter13_Lesson4.cpp
// Relationship of Observability to Sensor Placement
// Scratch implementation: matrix powers, observability matrix, and rank by Gaussian elimination.
// Compile: g++ -std=c++17 Chapter13_Lesson4.cpp -O2 -o Chapter13_Lesson4

#include <algorithm>
#include <cmath>
#include <iomanip>
#include <iostream>
#include <stdexcept>
#include <string>
#include <vector>

using Matrix = std::vector<std::vector<double>>;

Matrix eye(int n) {
    Matrix I(n, std::vector<double>(n, 0.0));
    for (int i = 0; i < n; ++i) I[i][i] = 1.0;
    return I;
}

Matrix multiply(const Matrix& A, const Matrix& B) {
    int m = static_cast<int>(A.size());
    int p = static_cast<int>(A[0].size());
    int n = static_cast<int>(B[0].size());
    Matrix C(m, std::vector<double>(n, 0.0));
    for (int i = 0; i < m; ++i) {
        for (int k = 0; k < p; ++k) {
            for (int j = 0; j < n; ++j) {
                C[i][j] += A[i][k] * B[k][j];
            }
        }
    }
    return C;
}

Matrix stackVertical(const std::vector<Matrix>& blocks) {
    int cols = static_cast<int>(blocks[0][0].size());
    Matrix out;
    for (const auto& B : blocks) {
        for (const auto& row : B) {
            if (static_cast<int>(row.size()) != cols) {
                throw std::runtime_error("Inconsistent column count.");
            }
            out.push_back(row);
        }
    }
    return out;
}

Matrix observabilityMatrix(const Matrix& A, const Matrix& C) {
    int n = static_cast<int>(A.size());
    Matrix Ak = eye(n);
    std::vector<Matrix> blocks;
    for (int k = 0; k < n; ++k) {
        blocks.push_back(multiply(C, Ak));
        Ak = multiply(Ak, A);
    }
    return stackVertical(blocks);
}

int rankMatrix(Matrix M, double tol = 1e-9) {
    int rows = static_cast<int>(M.size());
    int cols = static_cast<int>(M[0].size());
    int r = 0;
    for (int c = 0; c < cols && r < rows; ++c) {
        int pivot = r;
        for (int i = r + 1; i < rows; ++i) {
            if (std::fabs(M[i][c]) > std::fabs(M[pivot][c])) pivot = i;
        }
        if (std::fabs(M[pivot][c]) <= tol) continue;
        std::swap(M[r], M[pivot]);
        double piv = M[r][c];
        for (int j = c; j < cols; ++j) M[r][j] /= piv;
        for (int i = 0; i < rows; ++i) {
            if (i == r) continue;
            double factor = M[i][c];
            for (int j = c; j < cols; ++j) M[i][j] -= factor * M[r][j];
        }
        ++r;
    }
    return r;
}

Matrix CFromMask(int mask, int n) {
    Matrix C;
    for (int j = 0; j < n; ++j) {
        if (mask & (1 << j)) {
            std::vector<double> row(n, 0.0);
            row[j] = 1.0;
            C.push_back(row);
        }
    }
    return C;
}

int sensorCount(int mask) {
    int count = 0;
    while (mask) {
        count += mask & 1;
        mask >>= 1;
    }
    return count;
}

int main() {
    Matrix A = {
        {0.0, 1.0, 0.0, 0.0},
        {-2.0, -0.4, 0.8, 0.0},
        {0.0, 0.0, -1.0, 1.0},
        {0.6, 0.0, -3.0, -0.5}
    };

    int n = static_cast<int>(A.size());
    std::cout << "Sensor placement by rank of O = [C; C A; ...; C A^(n-1)]\n";

    for (int mask = 1; mask < (1 << n); ++mask) {
        if (sensorCount(mask) > 2) continue;
        Matrix C = CFromMask(mask, n);
        Matrix O = observabilityMatrix(A, C);
        int r = rankMatrix(O);
        std::cout << "Sensors {";
        bool first = true;
        for (int j = 0; j < n; ++j) {
            if (mask & (1 << j)) {
                if (!first) std::cout << ",";
                std::cout << (j + 1);
                first = false;
            }
        }
        std::cout << "} rank = " << r << "\n";
    }

    return 0;
}

File: Tutorials/Codes/Modern_Control_Course/Chapter13/Lesson4/Chapter13_Lesson4.java
// Chapter13_Lesson4.java
// Relationship of Observability to Sensor Placement
// Scratch implementation: observability matrix and numerical rank.
// Compile: javac Chapter13_Lesson4.java
// Run:     java Chapter13_Lesson4

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Chapter13_Lesson4 {
    static double[][] eye(int n) {
        double[][] I = new double[n][n];
        for (int i = 0; i < n; i++) I[i][i] = 1.0;
        return I;
    }

    static double[][] multiply(double[][] A, double[][] B) {
        int m = A.length;
        int p = A[0].length;
        int n = B[0].length;
        double[][] C = new double[m][n];
        for (int i = 0; i < m; i++) {
            for (int k = 0; k < p; k++) {
                for (int j = 0; j < n; j++) {
                    C[i][j] += A[i][k] * B[k][j];
                }
            }
        }
        return C;
    }

    static double[][] stack(List<double[][]> blocks) {
        int cols = blocks.get(0)[0].length;
        int rows = 0;
        for (double[][] block : blocks) rows += block.length;
        double[][] out = new double[rows][cols];
        int r = 0;
        for (double[][] block : blocks) {
            for (double[] row : block) {
                out[r++] = Arrays.copyOf(row, cols);
            }
        }
        return out;
    }

    static double[][] observabilityMatrix(double[][] A, double[][] C) {
        int n = A.length;
        double[][] Ak = eye(n);
        List<double[][]> blocks = new ArrayList<>();
        for (int k = 0; k < n; k++) {
            blocks.add(multiply(C, Ak));
            Ak = multiply(Ak, A);
        }
        return stack(blocks);
    }

    static int rank(double[][] input, double tol) {
        int rows = input.length;
        int cols = input[0].length;
        double[][] M = new double[rows][cols];
        for (int i = 0; i < rows; i++) M[i] = Arrays.copyOf(input[i], cols);

        int r = 0;
        for (int c = 0; c < cols && r < rows; c++) {
            int pivot = r;
            for (int i = r + 1; i < rows; i++) {
                if (Math.abs(M[i][c]) > Math.abs(M[pivot][c])) pivot = i;
            }
            if (Math.abs(M[pivot][c]) <= tol) continue;

            double[] temp = M[r];
            M[r] = M[pivot];
            M[pivot] = temp;

            double piv = M[r][c];
            for (int j = c; j < cols; j++) M[r][j] /= piv;

            for (int i = 0; i < rows; i++) {
                if (i == r) continue;
                double factor = M[i][c];
                for (int j = c; j < cols; j++) M[i][j] -= factor * M[r][j];
            }
            r++;
        }
        return r;
    }

    static double[][] cFromMask(int mask, int n) {
        int count = Integer.bitCount(mask);
        double[][] C = new double[count][n];
        int row = 0;
        for (int j = 0; j < n; j++) {
            if ((mask & (1 << j)) != 0) {
                C[row][j] = 1.0;
                row++;
            }
        }
        return C;
    }

    public static void main(String[] args) {
        double[][] A = {
            {0.0, 1.0, 0.0, 0.0},
            {-2.0, -0.4, 0.8, 0.0},
            {0.0, 0.0, -1.0, 1.0},
            {0.6, 0.0, -3.0, -0.5}
        };

        int n = A.length;
        System.out.println("Sensor placement by rank of O = [C; C A; ...; C A^(n-1)]");

        for (int mask = 1; mask < (1 << n); mask++) {
            if (Integer.bitCount(mask) > 2) continue;
            double[][] C = cFromMask(mask, n);
            double[][] O = observabilityMatrix(A, C);
            int r = rank(O, 1e-9);

            System.out.print("Sensors {");
            boolean first = true;
            for (int j = 0; j < n; j++) {
                if ((mask & (1 << j)) != 0) {
                    if (!first) System.out.print(",");
                    System.out.print(j + 1);
                    first = false;
                }
            }
            System.out.println("} rank = " + r);
        }
    }
}
